### PR TITLE
core: let the ACL decide if an user can edit their own person record

### PR DESCRIPTION
**Pull request by mworrell**
### Description

In the `z_acl` module it was hard coded that an user can always update their self.

This should be decided by the ACL module, as not all use cases allow an user to fully edit their own person resource.

### Checklist

- [ ] documentation updated
- [ ] tests added
- [ ] no BC breaks

### CODE REVIEW

File: apps/zotonic_core/src/support/z_acl.erl [hunk 1]
```diff
@@ -160,7 +160,7 @@ rsc_visible(RscName, Context) ->
 -spec rsc_prop_visible(m_rsc:resource(), atom() | binary(), z:context()) -> boolean().
 rsc_prop_visible(undefined, _Property, _Context) ->
     true;
-rsc_prop_visible(Id, _Property, #context{user_id=UserId}) when Id == UserId andalso is_integer(UserId) ->
+rsc_prop_visible(Id, _Property, #context{user_id=UserId}) when Id =:= UserId andalso is_integer(UserId) ->
```
Comment (vkatsuba) [inline apps/zotonic_core/src/support/z_acl.erl:163]:
```suggestion
rsc_prop_visible(Id, _Property, #context{ user_id=Id }) when is_integer(Id) ->
```

**Review by vkatsuba — APPROVED** [review]

**Review by vkatsuba — APPROVED** [review]

File: apps/zotonic_mod_acl_user_groups/src/support/acl_user_groups_checks.erl [hunk 2]
```diff
@@ -202,6 +202,14 @@ acl_is_allowed(_, #context{user_id=1}) ->
     true;
 acl_is_allowed(#acl_is_allowed{object=undefined}, _Context) ->
     undefined;
+acl_is_allowed(#acl_is_allowed{action=view, object=Id}, #context{ user_id = UserId })
+    when is_integer(Id), Id =:= UserId ->
```
Comment (vkatsuba) [inline apps/zotonic_mod_acl_user_groups/src/support/acl_user_groups_checks.erl:206]:
```suggestion
    when is_integer(Id) ->
```

File: apps/zotonic_mod_acl_user_groups/src/support/acl_user_groups_checks.erl [hunk 3]
```diff
@@ -202,6 +202,14 @@ acl_is_allowed(_, #context{user_id=1}) ->
...
+    % User can view their own resource
+    true;
+acl_is_allowed(#acl_is_allowed{action=update, object=Id}, #context{ user_id = UserId })
+    when is_integer(Id), Id =:= UserId ->
```
Comment (vkatsuba) [inline apps/zotonic_mod_acl_user_groups/src/support/acl_user_groups_checks.erl:210]:
```suggestion
    when is_integer(Id) ->
```

File: apps/zotonic_mod_acl_user_groups/src/support/acl_user_groups_checks.erl [hunk 4]
```diff
@@ -202,6 +202,14 @@ acl_is_allowed(_, #context{user_id=1}) ->
     true;
 acl_is_allowed(#acl_is_allowed{object=undefined}, _Context) ->
     undefined;
+acl_is_allowed(#acl_is_allowed{action=view, object=Id}, #context{ user_id = UserId })
```
Comment (vkatsuba) [inline apps/zotonic_mod_acl_user_groups/src/support/acl_user_groups_checks.erl:205]:
```suggestion
acl_is_allowed(#acl_is_allowed{ action=view, object=Id }, #context{ user_id = UserId })
```

File: apps/zotonic_mod_acl_user_groups/src/support/acl_user_groups_checks.erl [hunk 5]
```diff
@@ -202,6 +202,14 @@ acl_is_allowed(_, #context{user_id=1}) ->
...
+    when is_integer(Id), Id =:= UserId ->
+    % User can view their own resource
+    true;
+acl_is_allowed(#acl_is_allowed{action=update, object=Id}, #context{ user_id = UserId })
```
Comment (vkatsuba) [inline apps/zotonic_mod_acl_user_groups/src/support/acl_user_groups_checks.erl:209]:
```suggestion
acl_is_allowed(#acl_is_allowed{ action=update, object=Id }, #context{ user_id = UserId })
```

**Review by vkatsuba — APPROVED** [review]

File: apps/zotonic_mod_acl_user_groups/src/support/acl_user_groups_checks.erl (hunk 4 above)
Comment (mworrell) [reply apps/zotonic_mod_acl_user_groups/src/support/acl_user_groups_checks.erl]: Maybe we should re-do all clauses in this file, to make them consistent.

Comment (vkatsuba) [reply apps/zotonic_mod_acl_user_groups/src/support/acl_user_groups_checks.erl]: Yes, maybe also make sense to discuss plugins for formatting like https://github.com/AdRoll/rebar3_format

Comment (mworrell) [reply apps/zotonic_mod_acl_user_groups/src/support/acl_user_groups_checks.erl]: Yes, that would be a good thing to check and then have a one-time big update.

**Comment by mworrell** [comment]
Discussed with @robvandenbogaard, conclusion: also remove the user-view shortcuts, and depend on the ACL module to handle those.

File: apps/zotonic_mod_admin_identity/priv/templates/_admin_edit_sidebar_person_live.tpl [hunk 6]
```diff
@@ -1,4 +1,4 @@
-{% if m.acl.is_allowed.use.mod_admin_identity or id == m.acl.user %}
+{% if m.acl.is_allowed.use.mod_admin_identity or (id == m.acl.user and id.is_editable) %}
     <div class="form-group">
```
[the remark this answers is not in the document]
Comment (mworrell) [reply apps/zotonic_mod_admin_identity/priv/templates/_admin_edit_sidebar_person_live.tpl:2]: Idea is to only show this button to admins and the user themselves.

So you don't see this for a random other user you can edit.

**Comment by mmzeeman** [comment]
The acl_user_groups module has an option to logon (as admin), get all your rights, except allowing mutations. This is also hard coded in `z_acl`. That should probably also hold for allowing changing user credentials.

https://github.com/zotonic/zotonic/blob/345dc88926f9fd2fa419d434d10b1ce56245110b/apps/zotonic_mod_acl_user_groups/src/support/acl_user_groups_checks.erl#L366

**Comment by mworrell** [comment]
Maybe we should force a check on `acl_is_read_only` at the top of the pattern matches, to block all updates (except sudo.... that is another story)

**Review by mmzeeman — APPROVED** [review]

File: apps/zotonic_core/src/support/z_acl.erl [hunk 7]
```diff
@@ -192,10 +192,7 @@ rsc_prop_visible(RscName, Property, Context) ->
 -spec rsc_editable(m_rsc:resource(), z:context()) -> boolean().
 rsc_editable(undefined, _Context) ->
     false;
-rsc_editable(Id, #context{user_id=Id}) when is_integer(Id) ->
-    %% Can always edit myself
-    true;
-rsc_editable(_Id, #context{acl=admin}) ->
+rsc_editable(_Id, #context{ acl = admin }) ->
```
[the remark this answers is not in the document]
Comment (mworrell) [reply apps/zotonic_core/src/support/z_acl.erl:191]: If the `admin` is set then we are in a `sudo`, it is not when the admin user is logged on.

**Pull request merged**